plugins/woocommerce/src/Caches/OrderAggregateCache.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+declare( strict_types=1 );
+
+namespace Automattic\WooCommerce\Caches;
+
+use Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore;
+
+/**
+ * A class to cache aggregates from orders.
+ */
+class OrderAggregateCache {
+
+	/**
+	 * Option name.
+	 */
+	const OPTION_NAME = 'woocommerce_orders_aggregate_cache';
+
+	/**
+	 * The running totals for all order statuses.
+	 *
+	 * @var array|null
+	 */
+	private $totals = null;
+
+	/**
+	 * Whether or not the count has changed since loading.
+	 *
+	 * @var bool
+	 */
+	private $has_changed = false;
+
+	/**
+	 * Get the aggregate by identifier.
+	 *
+	 * @param string|string[] $status The order status to count.
+	 * @return int
+	 * @throws \Exception Exception thrown if status it not valid.
+	 */
+	public function get_count( $status ): int {
+		$status         = (array) $status;
+		$valid_statuses = $this->get_valid_statuses();
+
+		if ( ! empty( array_diff( $status, $valid_statuses ) ) ) {
+			throw new \Exception();
+		}
+
+		if ( ! $this->totals ) {
+			$cache        = get_option( self::OPTION_NAME );
+			$this->totals = $cache ? $cache : $this->count_orders_by_status();
+		}
+
+		return array_sum( array_intersect_key( $this->totals, array_flip( $status ) ) );
+	}
+
+	/**
+	 * Get valid statuses.
+	 *
+	 * @return string[]
+	 */
+	private function get_valid_statuses() {
+		return array_merge(
+			array_keys( wc_get_order_statuses() ),
+			array(
+				'trash',
+				'draft',
+				'auto-draft',
+			)
+		);
+	}
+
+	/**
+	 * Hydrate the initial cache.
+	 *
+	 * @return array
+	 */
+	private function count_orders_by_status() {
+		global $wpdb;
+
+		$orders_table = OrdersTableDataStore::get_orders_table_name();
+
+		$res = $wpdb->get_results(
+			"SELECT status, COUNT(*) AS cnt FROM {$orders_table} WHERE type = 'shop_order' GROUP BY status", // phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared
```

Comment:
@joshuatf I noticed this query is locked to `shop_order` type but I'm wondering if we could refactor this query somehow to support other custom order types i.e. `shop_order_refund` and `shop_subscription`?

Or is it intended that we build our own cache for subscriptions count?

---

One approach would be to update `$this->order_aggregate_cache->get_count()` to allow passing the order type as a param, then inside `count_orders_by_status` we could simply pass `$this->order_type` i.e. `$this->order_aggregate_cache->get_count( $status, $this->order_type );`

As for `OrdersTableQuery->get_count()` we could possibly pluck the order type from the `$where_clauses` and pass that to `$this->order_aggregate_cache->get_count()`.
Thoughts?

Reply:
Gah! Sorry Josh, I just saw your comment on the [related subscriptions issue](https://github.com/woocommerce/woocommerce-subscriptions/issues/4785) where you acknowledged this already:

> Currently caching is only being done for order types in the linked PR, but we may extend this to more order types.

Reply:
No problem, @mattallan!  I'll look into making this work for multiple order types.

Reply:
@mattallan, this should now work for all valid order types.  `OrderUtil::get_count_for_type( 'your-order-type' )`.

Please let me know if this works okay for your use case.